### ISSUE TITLE
Intermittent error when running a QE and adding to the Combat Tracker

### ISSUE THREAD

**Issue by spetzel2020** [spetzel2020/quick-encounters#119]
UPDATED:

Foundry 10.291, QE 1.1.5
```
foundry.js:46703 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'clear')
    at Token5e._refreshTarget (foundry.js:46703:17)
    at Token5e._refresh (foundry.js:46596:10)
    at Token5e.refresh (foundry.js:25224:10)
    at Canvas.highlightObjects (foundry.js:24094:11)
    at DrawingsLayer.deactivate (foundry.js:30469:12)
    at TilesLayer.activate (foundry.js:30438:76)
    at QuickEncounter.run (QuickEncounter.js:1214:30)
_refreshTarget @ foundry.js:46703
_refresh @ foundry.js:46596
refresh @ foundry.js:25224
highlightObjects @ foundry.js:24094
deactivate @ foundry.js:30469
activate @ foundry.js:30438
run @ QuickEncounter.js:1214
await in run (async)
(anonymous) @ QESheet.js:126
dispatch @ jquery.min.js:2
y.handle @ jquery.min.js:2
foundry.js:33189 Uncaught TypeError: Cannot read properties of undefined (reading 'width')
    at #drawTokenOcclusion (foundry.js:33189:37)
    at CanvasOcclusionMask.updateOcclusion (foundry.js:33167:29)
    at #update (foundry.js:50581:54)
    at t.emit (ticker.min.mjs:8:622)
    at e.update (ticker.min.mjs:8:3473)
    at _tick (ticker.min.mjs:8:1482)
#drawTokenOcclusion @ foundry.js:33189
updateOcclusion @ foundry.js:33167
#update @ foundry.js:50581
t.emit @ ticker.min.mjs:8
e.update @ ticker.min.mjs:8
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
requestAnimationFrame (async)
_tick @ ticker.min.mjs:8
foundry.js:46703 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'clear')
    at Token5e._refreshTarget (foundry.js:46703:17)
    at Token5e._refresh (foundry.js:46596:10)
    at Token5e.refresh (foundry.js:25224:10)
    at Token5e._onControl (foundry.js:47638:10)
    at Token5e.control (foundry.js:25319:10)
    at QuickEncounter.createCombat (QuickEncounter.js:1580:29)
    at QuickEncounter.js:1208:28
```

ORIGINAL REPORT
Foundry v10, QE 1.1.1

```
foundry.js:46703 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'clear')
    at Token5e._refreshTarget (foundry.js:46703:17)
    at Token5e._refresh (foundry.js:46596:10)
    at Token5e.refresh (foundry.js:25224:10)
    at Canvas.highlightObjects (foundry.js:24094:11)
    at DrawingsLayer.deactivate (foundry.js:30469:12)
    at TilesLayer.activate (foundry.js:30438:76)
    at QuickEncounter.run (QuickEncounter.js:1214:30)
foundry.js:46044 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'clear')
    at Token5e._refreshTarget (foundry.js:46044:17)
    at Token5e._refresh (foundry.js:45937:10)
    at Token5e.refresh (foundry.js:25134:10)
    at Token5e._onControl (foundry.js:46978:10)
    at Token5e.control (foundry.js:25229:10)
    at QuickEncounter.createCombat (QuickEncounter.js:1516:29)
    at QuickEncounter.js:1170:28
```

**Comment by spetzel2020** [comment]
Unable to reproduce; moving to 1.2

**Comment by spetzel2020** [comment]
From @dev7355608 in #dev-support: 
> Looks like the token is controlled before Token.draw has finished. Before that target is undefined and any refresh fails (control calls refresh)

**Comment by dev7355608** [comment]
The best solution is probably to take control of the token in the `createToken` hook.

```js
Hooks.on("createToken", (document) => {
  if ( !document.rendered) return;
  if ( /* ... */ ) document.object.control();
});
```

For reference `Token#_onCreate`:
```js
  /** @override */
  _onCreate(data, options, userId) {
    super._onCreate(data, options, userId);
    this.updateSource(); // Update vision and lighting sources
    if ( !game.user.isGM && this.isOwner && !this.document.hidden ) this.control({pan: true}); // Assume control
  }
```

**Comment by spetzel2020** [comment]
> From @dev7355608 in #dev-support:
> 
> > Looks like the token is controlled before Token.draw has finished. Before that target is undefined and any refresh fails (control calls refresh)
> 
> Try adding a timeout or check.

There is already a 200ms timeout between 1208: `createCombat()` and 1214:`canvas.tiles.activate()`

**Comment by spetzel2020** [comment]
In 10.303, you get only part of this error and it no longer seems to prevent combat:
```
foundry.js:46386 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'clear')
    at Token5e._refreshTarget (foundry.js:46386:17)
    at Token5e._refresh (foundry.js:46279:10)
    at Token5e.refresh (foundry.js:25283:10)
    at Canvas.highlightObjects (foundry.js:24153:11)
    at DrawingsLayer.deactivate (foundry.js:30528:12)
    at TilesLayer.activate (foundry.js:30497:76)
    at QuickEncounter.run (QuickEncounter.js:1214:30)
```

**Comment by spetzel2020** [comment]
Closing as fixed in 1.1.5 and Foundry 10.303
